### PR TITLE
[Tom] Throw exception if sentinel enum value attempted to be encoded.

### PR DESCRIPTION
**Pull request by tom-smalls**
Encoding the enum sentinel values is not something that should happen, and is usually the result of a programming error. Especially for strings, where the sentinel values can lead to a corrupt FIX message, you want to avoid this.

One thing we weren't sure about is whether to treat NULL_VALUE and ARTIO_UNKNOWN differently, silently dropping NULL_VALUE and throwing for ARTIO_UNKNOWN. Alternatively we could silently drop in both cases. What are your thoughts on this?

### CODE REVIEW

**Comment by RichardWarburton** [comment]
I think silently dropping NULL_VALUE and throwing for ARTIO_UNKNOWN makes most sense to me. Essentially setting the value for `field` to NULL_VALUE would have the same effect as calling `resetField()`.

File: artio-codecs/src/test/java/uk/co/real_logic/artio/dictionary/generation/EncoderGeneratorTest.java [hunk 1]
```diff
@@ -1029,4 +1075,21 @@ private void setDataFieldLength(final Encoder encoder) throws Exception
     {
         setInt(encoder, "dataFieldLength", 3);
     }
+
+    private static void assertThrows(final ThrowingRunnable runnable, final Class<? extends Exception> thrown)
+    {
```
Comment (RichardWarburton) [inline artio-codecs/src/test/java/uk/co/real_logic/artio/dictionary/generation/EncoderGeneratorTest.java]: Maybe I'm misreading this, but it looks like this assert would pass if no exception is thrown and only fail if an exception of the throw type is thrown?

Wouldn't it be better to use [Junit's built in support for this](https://junit.org/junit4/javadoc/4.12/org/junit/rules/ExpectedException.html)

**Comment by tom-smalls** [comment]
Sure

File: artio-codecs/src/test/java/uk/co/real_logic/artio/dictionary/generation/EncoderGeneratorTest.java (hunk 1 above)
Comment (tom-smalls) [reply artio-codecs/src/test/java/uk/co/real_logic/artio/dictionary/generation/EncoderGeneratorTest.java]: Yes, I forgot about that functionality; we use JUnit5, which has a different way for asserting on exceptions and could only remember the @Test(Exception.class) way.

I've fixed this up

**Comment by tom-smalls** [comment]
I have made the changes to just drop for NULL_VAL

**Comment by RichardWarburton** [comment]
I was about to merge this PR, but your other PR, that I just merged, deletes setEnum() that this PR uses.

**Comment by tom-smalls** [comment]
Ahhh, ok. Will fix now

**Comment by tom-smalls** [comment]
I've updated the PR to remove the merge conflicts with the other changes.

**Pull request merged**